upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+func (m *ManagedDatabaseMetricsChartFloat64) UnmarshalJSON(d []byte) error {
+	var chart canonicalManagedDatabaseMetricChartForUnmarshal
+	if err := json.Unmarshal(d, &chart); err != nil {
+		return err
+	}
+	timestamps := make([]time.Time, 0, len(chart.Data.Rows))
+	rows := make([][]float64, 0, len(chart.Data.Rows))
+	inRows := chart.Data.Rows
+	for i := range inRows {
+		var row = make([]float64, 0, len(inRows)-1)
+		var ts time.Time
+		if len(inRows[i]) != len(chart.Data.Columns) {
+			return fmt.Errorf("unexpected number of columns at row %d (not %d)", i, len(m.Columns))
+		}
+		if err := json.Unmarshal(inRows[i][0], &ts); err != nil {
+			return fmt.Errorf("cannot unmarshal timestamp at row %d: %w", i, err)
+		}
+		timestamps = append(timestamps, ts)
+		for j := 1; j < len(inRows[i]); j++ {
+			var iv float64
+			if err := json.Unmarshal(inRows[i][j], &iv); err != nil {
+				return fmt.Errorf("cannot unmarshal value at row %d, column %d: %w", i, j, err)
+			}
+			row = append(row, iv)
+		}
+		rows = append(rows, row)
+	}
+	m.Timestamps = timestamps
+	m.Columns = chart.Data.Columns[1:]
+	m.Title = chart.Hints.Title
+	m.Rows = rows
+	return nil
+}
+
+// ManagedDatabaseMetricsChartHeader represents common fields of a metrics chart
+type ManagedDatabaseMetricsChartHeader struct {
+	// Columns contains a set of columns that describe for what node the corresponding row element belongs to
+	// as well as the type of the metric value
+	Columns []ManagedDatabaseMetricsColumn
+	// Timestamps contains the timestamps of the rows. Its indexing corresponds the rows.
+	Timestamps []time.Time
+	// Title contains a description of the metrics chart
+	Title string
+}
+
+type canonicalManagedDatabaseMetricChartForMarshal struct {
+	Data struct {
+		Columns []ManagedDatabaseMetricsColumn `json:"cols"`
+		Rows    [][]interface{}                `json:"rows"`
+	} `json:"data"`
+	Hints struct {
+		Title string `json:"title"`
+	} `json:"hints"`
+}
+
+type canonicalManagedDatabaseMetricChartForUnmarshal struct {
+	Data struct {
+		Columns []ManagedDatabaseMetricsColumn `json:"cols"`
+		Rows    [][]json.RawMessage            `json:"rows"`
+	} `json:"data"`
+	Hints struct {
+		Title string `json:"title"`
+	} `json:"hints"`
+}
+
+// ManagedDatabaseMetricsChartInt represents a metric chart with int row values
+type ManagedDatabaseMetricsChartInt struct {
+	ManagedDatabaseMetricsChartHeader
+	// Rows contains a slice of values per row. The inner slice has the same indexing as the
+	// ManagedDatabaseMetricsChartHeader.Columns
+	Rows [][]int
+}
+
+// MarshalJSON implements json.Marshaler for the canonical form of metrics chart
+func (m ManagedDatabaseMetricsChartInt) MarshalJSON() ([]byte, error) {
+	var chart canonicalManagedDatabaseMetricChartForMarshal
+	chart.Data.Columns = make([]ManagedDatabaseMetricsColumn, 0, len(m.Columns)+1)
+	chart.Data.Columns = []ManagedDatabaseMetricsColumn{{Label: "time", Type: "date"}}
```

Comment:
Same as above, should probably be append() rather than creating a new slice?

Reply:
implemented